Mount injection

LizaTretyakova-zz opened the pull request:
Just for the ease of review

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,131 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+	// Do it here to be able to use temp in both parent and child
+	pref = getenv("LXC_DIR");
+	if (pref == NULL)
+		pref = "/var/lib/lxc";
```
We don't support `LXC_DIR` for `LXC` as we do `LXD_DIR` for `LXD`. :)

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,131 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+	if (ret < 0 || ret >= len) {
+		ERROR("Error writing shmounts tempdir name");
+		return -1;
+	}
```
The path is wrong. You do not want to create a temporary mountpoint on the host at `/var/lib/lxc/shmounts/<container-name>/<tmp>` but at `conf->shmounts_path/<tmp>`. :)

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,131 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+		// Do the mount
+		ret = mount(source, temp, filesystemtype, mountflags, data);
+		if (ret < 0)
+			return -1;
```
This mount needs to be removed on error or after we have moved it in place in the container. :)

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,131 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+			return -1;
+		// if fully unprivileged
+		if (am_guest_unpriv() && !lxc_list_empty(&c->lxc_conf->id_map)) {
+			ret = setns(init_pid,CLONE_NEWUSER);
```
You need to attach to the user namespace first and then to the mount namespace. Otherwise the kernel will refuse the `setns()` to the mount namespace.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,131 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+		if (!suff)
+			return -1;
+
+		len = strlen(c->lxc_conf->shmounts_path) + strlen(suff);
```
`shmounts_path` will refer to the host's side though.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,131 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+
+	ret = unlink(temp);
+	if (ret < 0)
+		return -1;
```
Ah, we need to improve the error handling. We might just do:

    (void)umount2();
    (void)unlink();

and don't care about the error.

brauner replied:
You don't need to duplicate the error handling. This can be unified by using a very common `C` code pattern which you might not yet be familiar with. Something like:

```
foo()
{
        /* global return value for this function */
        int fret = -1;

        /* some error happens */
        if (error)
                goto on_error:


        /* function is successful so initialize the global return value to indicate success */
        fret = 0;
        /* you can simply fallthrough now */

        on_error:
                (void)umount2(temp, MNT_DETACH);
                (void)unlink(temp);

        return fret;
}
```

brauner commented on src/lxc/conf.h:
```diff
@@ -365,6 +365,9 @@ struct lxc_conf {
 	struct lxc_list procs;
+
+	/* path to the shared mounpoint */
+	char *shmounts_path;
```
So I think you want two paths here:

    /* host side of shared mountpoint */
    char *shmounts_host;

    /* container side of shared mountpoint */
    char *shmounts_cont;

The use will give us `shmounts_path` via `lxc.mount.auto = inject:/shmounts/path` and we will internally figure out the `shmounts_cont` path, e.g. `/dev/.lxc-mounts`.

Might also want to make this a struct if you want:

    struct lxc_shmount {
            char *path_host;
            char *path_cont;
    };

That should also clarify the logic in `c->mount()` a little more.

brauner reviewed (commented):
Heading in the right direction. :)

brauner commented on src/lxc/conf.c:
```diff
@@ -3855,6 +3855,7 @@ void lxc_conf_free(struct lxc_conf *conf)
 	lxc_clear_procs(conf, "lxc.proc");
 	free(conf->cgroup_meta.dir);
 	free(conf->cgroup_meta.controllers);
+	free(conf->lxc_shmount.path_host);
```
missing a

    +	free(conf->lxc_shmount.path_cont);

brauner commented on src/lxc/conf.h:
```diff
@@ -185,6 +185,9 @@ enum {
 	LXC_AUTO_CGROUP_MASK          = 0x1F0,   /* all known cgroup options, doe not contain LXC_AUTO_CGROUP_FORCE */
+
+	LXC_AUTO_SHMOUNTS             = 0x200,   /* shared mount point */
+	LXC_AUTO_SHMOUNTS_MASK        = 0x200,   /* shared mount point mask */
```
Ugh, this enum is annoying to think through. We might have to revisit those values later.

LizaTretyakova-zz replied:
Okay. I just found and followed this comment https://github.com/LizaTretyakova/lxc/pull/1/files#diff-2f3344d3cb9700d11140decc560aaacaL1746 although I don't know about the future plans regarding this enum.

brauner replied:
Yeah, this is not your fault. The enum is just setup in a very specific way and I always have to do the math to get it right. But we'll defer this.

brauner commented on src/lxc/conf.h:
```diff
@@ -365,6 +368,13 @@ struct lxc_conf {
 	/* procs */
 	struct lxc_list procs;
+
+	struct {
```
Hm, make it:
```
struct lxc_shmount {
        /* stuff */
} lxc_shmount;
```

brauner commented on src/lxc/confile.c:
```diff
@@ -1760,6 +1761,8 @@ static int set_config_mount_auto(const char *key, const char *value,
 		return -1;
 
 	for (autoptr = autos;; autoptr = NULL) {
+		bool is_shmounts = false;
```
Nit: Should be `is_shmount` since its only one. :)

brauner replied:
Actually, just ignore this.

brauner commented on src/lxc/confile.c:
```diff
@@ -1769,6 +1772,12 @@ static int set_config_mount_auto(const char *key, const char *value,
+
+			if (!strcmp("shmounts:", allowed_auto_mounts[i].token) &&
+					!strncmp("shmounts:", token, strlen("shmounts:"))) {
+				is_shmounts = true;
```
Hm, that check looks wrong. It should be:

```
if (strncmp("shmounts:", token, sizeof("shmounts:") - 1) == 0) {
        is_shmounts = true;
        break;
}
```

LizaTretyakova-zz replied:
Hm. I think that we need to check the `allowed_auto_mounts[i].token` as well in order to yield the proper index `i`? Or you meant the `sizeof` vs `strlen`?

LizaTretyakova-zz replied:
@brauner

brauner commented on src/lxc/confile.c:
```diff
@@ -1778,6 +1787,10 @@ static int set_config_mount_auto(const char *key, const char *value,
 		lxc_conf->auto_mounts |= allowed_auto_mounts[i].flag;
+		if (is_shmounts) {
+			lxc_conf->lxc_shmount.path_host = strdup(token + strlen("shmounts:"));
+			lxc_conf->lxc_shmount.path_cont = "/dev/.lxc-mounts";
```
Hm, `strdup("/dev/.lxc-mounts")` right away so we don't get into a situation where we accidently free static memory. :)

LizaTretyakova-zz replied:
Yeah, right

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,159 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+	char *path;
+	pid_t pid;
+	pid_t init_pid;
+	struct stat sb;
```
```
int fd, len, ret, status;
char *path, *suff, *temp;
pid_t init_pid, pid;
struct stat sb;
```

brauner replied:
Nit: `temp` should traditionally rather be `tmp`. :)

LizaTretyakova-zz replied:
Well, it was actually intended to be `template`... Well, not that it meant that I shouldn't rename it :)

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,159 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+	struct stat sb;
+
+	len = strlen(c->lxc_conf->lxc_shmount.path_host) + strlen("/lxcmountXXXXXX") + 4;
+	temp = (char *)malloc(len);
```
We don't cast the return value of `malloc()`. I know, some people argue that this can lead to bugs e.g. if you:

```
struct bla *a;
struct blu *b;

b = malloc(sizeof(struct bla));
```
then the `C` standard mandates an implicit conversion of `void *` that `malloc()` returns to the type on the left side so you might end up with undefined behavior. We're still not doing it though, the kernel isn't also, glibc also isn't doing it. :)

LizaTretyakova replied:
So am I right that the logic is "everyone is doing the unsafe behaviour so will we"? :) On a serious note, sure, if this is the conventional codestyle.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,159 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+	pid_t init_pid;
+	struct stat sb;
+
+	len = strlen(c->lxc_conf->lxc_shmount.path_host) + strlen("/lxcmountXXXXXX") + 4;
```
I usually prefer `sizeof() - 1` for static strings such as `/lxcmountXXXXXX` even though the compiler should optimize at the appropriate level but still.
If you do stuff like `+4` in `C`, please always justify in a comment what those `4` extra spaces are for. This makes it easier later to reason about what's going on. Examples are:
- https://github.com/lxc/lxc/blob/22b2b9c87cd63e6bd7445c2159a0f7d6aa2b7368/src/lxc/cgroups/cgfsng.c#L1796
- https://github.com/lxc/lxc/blob/22b2b9c87cd63e6bd7445c2159a0f7d6aa2b7368/src/lxc/conf.c#L1175

LizaTretyakova-zz replied:
Aand it was actually a mistake here...

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,159 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+	}
+
+	// Create a temporary dir / file
+	// **under the shared mountpoint**
```
We only use `/* */` style comments.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,159 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+			// if fully unprivileged
+			ret = setns(init_pid,CLONE_NEWUSER);
+			if (ret < 0) {
+				ERROR("Failed to join container's userns");
```
Should be `SYSERROR()` because it will automatically print the `errno`.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,159 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+		}
+		ret = setns(init_pid, CLONE_NEWNS);
+		if (ret < 0) {
+			ERROR("Failed to join container's mountns");
```
Should be `SYSERROR()` because it will automatically print the `errno`.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,159 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+		// Do the mount
+		ret = mount(source, temp, filesystemtype, mountflags, data);
+		if (ret < 0) {
+			ERROR("Failed to mount");
```
Should be `SYSERROR()` because it will automatically print the `errno`.

brauner replied:
I usually will be descriptive like:

    SYSERROR("Failed to mount \"%s\" onto \"%s\", source, temp);

but that's not so important right now. We will have more reviews. :)

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,159 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+		if (!suff)
+			goto out_child_err_temp;
+
+		len = strlen(c->lxc_conf->lxc_shmount.path_cont) + strlen(suff);
```
Hm, don't you know the length of `suff`?  It should just be `sizeof("/lxcmountXXXXX") - 1` or something like this.

LizaTretyakova-zz replied:
Yeah, right, we can do so. Though I am a bit worried about the static constants. I saw that we use them in the code but they are usually kept in one place. And here there is almost a screen of code between the occurrences. Is it fine?

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,159 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+			goto out_child_err_temp;
+
+		len = strlen(c->lxc_conf->lxc_shmount.path_cont) + strlen(suff);
+		path = (char *)malloc(len);
```
Same, please don't cast the return value of `malloc()`.

brauner reviewed (commented):
Left a few more comments. :)

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,152 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+			const char *source, const char *target,
+			const char *filesystemtype, unsigned long mountflags,
+			const void *data, struct lxc_mount *mnt) {
+	int ret = -1, len, fd, status;
```
Please separate initialized from uninitialized variables. :)
See https://github.com/lxc/lxc/blob/4a308d6693b3246e5e22cc57ed615d39afda818b/CODING_STYLE.md#declaring-variables

LizaTretyakova-zz replied:
Oooooww... My bad. I should have read it way earlier =\

brauner replied:
No worries :)

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,152 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+	struct stat sb;
+
+	len = strlen(c->lxc_conf->lxc_shmount.path_host) + (sizeof("/lxcmountXXXXXX") - 1);
+	template = malloc(len);
```
You need `malloc(len + 1)` for the terminating `\0` byte. This also ensures that the check
    
    ret = snprintf();
    if (ret < 0 || ret >= len)

below makes sense.

LizaTretyakova-zz replied:
If I get everything right, just not subtracting 1 above should be enough (both times)

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,152 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+			const char *source, const char *target,
+			const char *filesystemtype, unsigned long mountflags,
+			const void *data, struct lxc_mount *mnt) {
+	int ret = -1, len, fd, status;
```
Use `size_t` instead of `int` for `len` since it is defined as the datatype to use when querying the size of an object. Also `malloc()` and friends expect `size_t`.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,152 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+		goto out_err_free;
+	}
+	ret = snprintf(template, len, "%s/lxcmountXXXXXX", c->lxc_conf->lxc_shmount.path_host);
+	if (ret < 0 || ret >= len) {
```
Since you will now have declared `len` as `size_t` you need to cast `ret` to `size_t` otherwise compilers will complain about comparing integers of different signs/size:

    if (ret < 0 || (size_t)ret >= len)

This cast is safe by the time you reach the right hand-side of the conjunction you know that `ret` is not negative and you know on principle grounds that `size_t` will hold a positive number at least as large as `int` can.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,152 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+		}
+	} else if (S_ISDIR(sb.st_mode)) {
+		template = mkdtemp(template);
+		if (template == NULL) {
```
Use `if (!template)` which is what we do everywhere else in the codebase. :)

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,152 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+		/* Enter the container namespaces */
+		if (am_guest_unpriv() && !lxc_list_empty(&c->lxc_conf->id_map)) {
+			/* if fully unprivileged */
+			ret = setns(init_pid,CLONE_NEWUSER);
```
Nit add a space after the comma between `init_pid,CLONE_NEWUSER`

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,152 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+			goto out_child_err_temp;
+
+		len = strlen(c->lxc_conf->lxc_shmount.path_cont) + (sizeof("/lxcmountXXXXXX") - 1);
+		path = malloc(len);
```
Same, you need `malloc(len + 1)`.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,152 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+			goto out_child_err_temp;
+		}
+		ret = snprintf(path, len, "%s%s", c->lxc_conf->lxc_shmount.path_cont, suff);
+		if (ret < 0 || ret >= len) {
```
Again, cast of `ret` to `size_t` is needed now that `len` is of type `size_t`.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,152 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+		goto out_err_free;
+	}
+	if (WIFEXITED(status) && WEXITSTATUS(status) != EXIT_SUCCESS)
+		goto out_err_unmount;
```
You should use our helper [wait_for_pid](https://github.com/lxc/lxc/blob/83ffaa1d39096ee2897c6f35d523a819698d1daa/src/lxc/utils.c#L278) which will do all of that for you.

brauner commented on src/lxc/confile.c:
```diff
@@ -1769,6 +1772,12 @@ static int set_config_mount_auto(const char *key, const char *value,
+
+			if (!strcmp("shmounts:", allowed_auto_mounts[i].token) &&
+					!strncmp("shmounts:", token, sizeof("shmounts:") - 1)) {
+				is_shmounts = true;
```
Oh now I see what you meant when you said
> Hm. I think that we need to check the allowed_auto_mounts[i].token as well in order to yield the proper index i? Or you meant the sizeof vs strcmp?

You're right. Good catch!

brauner reviewed (commented):
Some more comments. :)

LizaTretyakova-zz commented on src/lxc/lxccontainer.c:
```diff
@@ -4691,6 +4691,11 @@ static int do_lxcapi_mount(struct lxc_container *c,
 	struct stat sb;
 	int ret = -1;
 
+	if (!c || !c->lxc_conf) {
```
I thought that such check might be useful, to be sure that we do not dereference null

brauner replied:
Yep, makes sense.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,158 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+	}
+
+	len = strlen(c->lxc_conf->lxc_shmount.path_host) + sizeof("/lxcmountXXXXXX");
+	template = malloc(len);
```
Hm, no what I meant is:

```
len = strlen(c->lxc_conf->lxc_shmount.path_host) + sizeof("/lxcmountXXXXXX") - 1;
template = malloc(len + 1);
```
This way the right-hand side of the conjunct in the error check for `snprintf()` will make sense:
```
ret = snprintf(template, len, "%s/lxcmountXXXXXX", c->lxc_conf->lxc_shmount.path_host);
if (ret < 0 || (size_t)ret >= len)
```

LizaTretyakova-zz replied:
Ah, I finally see where I was mistaken. I saw the part of the `man snprintf` about writing `at most _size_ bytes (including the terminating null byte ('\0'))`, but I missed the `return the number of characters printed (excluding the null byte <...>)`.

LizaTretyakova-zz replied:
Hmm. Wait. 
In your code snippet, the length of the meaningful part of the string is `len`, and the length with the terminating null-byte is `len + 1`. `len + 1` bytes are `malloc`ed. Further, `snprintf` would write `<= len` bytes, and the null-byte would be assumed among these `len` (which would not be true).
The check, AFAIU, would make sense anyway as long as the parameter matches the one we passed to the `snprintf` function as __size__ .

brauner replied:
`len` is also used to tell `snprintf()` how many bytes it is supposed to write at maximum:

    snprintf(..., len, ...);

If `len` does include the count for the `\0` byte you may get truncated results although it didn't have to be that way.

brauner replied:
The generic pattern in the codebase is to calculate the length without the terminating zero byte since this is in line with what `strlen()` does. The terminating zero byte is only accounted for when you allocate a buffer.

brauner replied:
Please switch to the `malloc(len + 1)` style everywhere so that the length of a string means exactly what `strlen()` makes it mean: all bytes **excluding** the `\0` byte. :)

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,158 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+			goto out_child_err_temp;
+
+		len = strlen(c->lxc_conf->lxc_shmount.path_cont) + sizeof("/lxcmountXXXXXX");
+		path = malloc(len);
```
Same comment as above.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,158 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+	if (!c || !c->lxc_conf) {
+		ERROR("Container or configuration is NULL");
+		goto out_err;
+	}
```
Actually, there's no need to use a `goto` here since you don't need to cleanup a bunch of stuff. So you should just:

    return -EINVAL

here.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,158 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+	template = malloc(len);
+	if (!template) {
+		SYSERROR("Not enough memory");
+		goto out_err_free;
```
Same, you can just `return -ENOMEM` here since there's nothing to cleanup yet.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,158 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+	}
+
+	if (S_ISREG(sb.st_mode)) {
+		fd = mkstemp(template);
```
Seems you're leaking the `fd`. This needs to be closed when `fd >= 0`.
So at the top of the function do:

    int fd = -EBADF;

Then when you jump to `out_err_free` add a condition:

    if (fd >= 0)
            close(fd);

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,158 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+		ret = mount(source, template, filesystemtype, mountflags, data);
+		if (ret < 0) {
+			SYSERROR("Failed to mount \"%s\" onto \"%s\"", source, template);
+			goto out_child_err_temp;
```
`_exit(EXIT_FAILURE)` here directly since there's nothing to cleanup yet.

LizaTretyakova-zz replied:
I though that it's better to `free()` the `template` here since it was `malloc()`ed?

LizaTretyakova-zz replied:
(and the same logic was for below)

brauner replied:
Oh right, you're not using a static buffer. Makes sense.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,158 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+		init_pid = do_lxcapi_init_pid(c);
+		if (init_pid < 0) {
+			ERROR("Failed to obtain container's init pid");
+			goto out_child_err_temp;
```
`_exit(EXIT_FAILURE)` here directly since there's nothing to cleanup yet.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,158 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+			ret = setns(init_pid, CLONE_NEWUSER);
+			if (ret < 0) {
+				SYSERROR("Failed to join container's userns");
+				goto out_child_err_temp;
```
`_exit(EXIT_FAILURE)` here directly since there's nothing to cleanup yet.

brauner replied:
So about those `setns()` what you're doing now is invalid. If you do

    man setns

you'll see that the first argument of `setns()` is `int fd` which is a file descriptor but you're passing the `init_pid`. I remember that I jotted it down this way in the spec we wrote so I should've clarified this (and please update the spec accordingly :)). What you need to do is:

    int userns_fd = open("/proc/<init_pid>/ns/user", O_RDONLY | O_CLOEXEC);
    int mntns_fd = open("/proc/<init_pid>/ns/mnt", O_RDONLY | O_CLOEXEC);

and pass the correct `fd` to the corresponding `setns()`. Which also means you need to cleanup those `fd`s too later on. ;)

The good news about this is that it won't get more complex because we have a helper in `utils.{c,h}` that does all of this for you:

```
bool switch_to_ns(pid_t pid, const char *ns) {
	int fd, ret;
	char nspath[MAXPATHLEN];

	/* Switch to new ns */
	ret = snprintf(nspath, MAXPATHLEN, "/proc/%d/ns/%s", pid, ns);
	if (ret < 0 || ret >= MAXPATHLEN)
		return false;

	fd = open(nspath, O_RDONLY);
	if (fd < 0) {
		SYSERROR("failed to open %s", nspath);
		return false;
	}

	ret = setns(fd, 0);
	if (ret) {
		SYSERROR("failed to set process %d to %s of %d.", pid, ns, fd);
		close(fd);
		return false;
	}
	close(fd);
	return true;
}
```

LizaTretyakova-zz replied:
Fixed the spec (it was slight :)) 
(the code too)

brauner reviewed (changes requested):
Spotted a few more bugs. :)

LizaTretyakova-zz commented on src/lxc/lxccontainer.c:
```diff
@@ -4695,14 +4695,14 @@ static int do_lxcapi_mount(struct lxc_container *c,
 	}
-	ret = snprintf(template, len, "%s/lxcmountXXXXXX", c->lxc_conf->lxc_shmount.path_host);
-	if (ret < 0 || (size_t)ret >= len) {
+	ret = snprintf(template, len + 1, "%s/lxcmountXXXXXX", c->lxc_conf->lxc_shmount.path_host);
```
> If `len` does include the count for the `\0` byte you may get truncated results although it didn't have to be that way.

Why? It would be truncated if we ask for a too small limit, i.e. `len` not including `\0`? 

But I changed the `len` so it is now strlen-like :)

brauner replied:
Nice.

brauner commented on src/lxc/lxccontainer.c:
```diff
@@ -4659,6 +4680,156 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
+		return -ENOMEM;
+	}
+	ret = snprintf(template, len + 1, "%s/lxcmountXXXXXX", c->lxc_conf->lxc_shmount.path_host);
+	if (ret < 0 || (size_t)ret >= len + 1) {
```
Oh, I missed this :)

brauner commented on src/lxc/confile.c:
```diff
@@ -1778,6 +1787,15 @@ static int set_config_mount_auto(const char *key, const char *value,
+					strdup(token + (sizeof("shmounts:") - 1));
+			if (!strcmp(lxc_conf->lxc_shmount.path_host, "")) {
+				ERROR("Invalid shmounts path: empty");
+				break;
```
Hm, it would be better to verify that this isn't empty before the `strdup()`. You can use:

```
bool lxc_config_value_empty(const char *value)
{
	if (value && strlen(value) > 0)
		return false;

	return true;
}
```
This is the simple verification we need right now.

brauner reviewed (commented):
Looks good now.
Next step is to implement the logic to create the shared mountpoint on the host.

brauner commented on src/lxc/conf.c:
```diff
@@ -623,6 +623,37 @@ unsigned long add_required_remount_flags(const char *s, const char *d,
+	if (ret < 0 || ret >= len_mount + 1) {
+		ret = -1;
+		goto out_free_new_mount;
+	}
```
In such cases, where there's only one jump location I would suggest to use:
```
if (/* error */) {
        free(new_mount);
        return -1;
}
```
but you can leave it as is. :)

brauner commented on src/lxc/conf.c:
```diff
@@ -757,79 +788,80 @@ static int lxc_mount_auto_mounts(struct lxc_conf *conf, int flags, struct lxc_ha
 		}
 
+		/* Check if shmount point is already set up */
 		if(is_shared_mountpoint(conf->lxc_shmount.path_host)
```
nit: missing ` ` between `if` and `(`. :)

brauner commented on src/lxc/conf.c:
```diff
@@ -757,79 +788,80 @@ static int lxc_mount_auto_mounts(struct lxc_conf *conf, int flags, struct lxc_ha
-			goto out_free_path;
+				&& !stat(full_cont_path, &sb) && S_ISDIR(sb.st_mode)) {
+			/* The host path is a shared mounpoint
+			 * (TODO: not checked for MS_REC),
```
That comment can be removed. :)

brauner commented on src/lxc/conf.c:
```diff
@@ -757,79 +788,80 @@ static int lxc_mount_auto_mounts(struct lxc_conf *conf, int flags, struct lxc_ha
-		size_t len_mount;
+		size_t len_cont;
+		char *full_cont_path;
+		struct stat sb;
```
Move the declaration of `struct stat sb` into the `if (is_shared_...)` condition since you're only using it there. Keep the scope tight. :)

brauner commented on src/lxc/conf.c:
```diff
@@ -757,79 +788,80 @@ static int lxc_mount_auto_mounts(struct lxc_conf *conf, int flags, struct lxc_ha
-			/* Already set up */
-			ret = 0;
-			goto out_free_path;
+				&& !stat(full_cont_path, &sb) && S_ISDIR(sb.st_mode)) {
```
Hm, don't you need to call `stat()` on `path_host` and not on `full_cont_path`? How about:
```
if (is_shared_mountpoint(conf->lxc_shmount.path_host) {
        int ret;
        struct stat sb;

        ret = stat(conf->lxc_shmount.path_host, &sb);
        if (ret == 0 && S_ISDIR(sb.st_mode)) {
                INFO("Path \"%s\" is already MS_SHARED. Reusing", conf->lxc_shmount.path_host);
                goto finalize;
        }
}
```

LizaTretyakova-zz replied:
Oh, right, I see. But do we need this check at all then? `mount` seems to return `ENOTDIR` if `path_host` was not a directory

brauner replied:
Sure, if that simplifies things. :)

brauner commented on src/lxc/conf.c:
```diff
@@ -757,79 +788,80 @@ static int lxc_mount_auto_mounts(struct lxc_conf *conf, int flags, struct lxc_ha
+		/* Borrowed from mount_entry_create_dir_file()
+		 *
+		 * Create host and cont mount paths
+		 */
```
You can remove this comment. It's ok to copy and paste stuff from within our library without comments. :)

brauner commented on src/lxc/conf.c:
```diff
@@ -757,79 +788,80 @@ static int lxc_mount_auto_mounts(struct lxc_conf *conf, int flags, struct lxc_ha
+		 *
+		 * Create host and cont mount paths
+		 */
 		ret = mkdir_p(conf->lxc_shmount.path_host, 0755);
```
Hm, make that `0711`. It's safe and that's all the permissions we need.

brauner commented on src/lxc/conf.c:
```diff
@@ -757,79 +788,80 @@ static int lxc_mount_auto_mounts(struct lxc_conf *conf, int flags, struct lxc_ha
+			free(full_cont_path);
+			return ret;
+		}
+		ret = mkdir_p(full_cont_path, 0755);
```
Hm, make that 0711. It's safe and that's all the permissions we need.

brauner commented on src/lxc/conf.c:
```diff
@@ -757,79 +788,80 @@ static int lxc_mount_auto_mounts(struct lxc_conf *conf, int flags, struct lxc_ha
-			SYSERROR("Not enough memory");
-			goto out_free_path;
+finalize:
+		ret = add_shmount_to_conf(conf, full_cont_path);
```
You shouldn't and can't add `full_cont_path` here. What you should simply add is the relative target for the bind-mount in the container. So if the bind-mount target for the shared mountpoint in the container is `/dev/.lxc` then you should pass in:

    add_shmount_to_conf(conf, "dev/.lxc")

since LXC will automatically take relative targets to be relative to the container's rootfs. :)

brauner reviewed (changes requested):
Ok, overall looks good. Just a few changes.

brauner commented on src/tests/mount_injection.c:
```diff
@@ -0,0 +1,279 @@
+		goto out_pipe;
+	}
+	attach_options.stdout_fd = pipefd[1];
+	ret = c->attach(c, find_in_proc_mounts, d, &attach_options, &pid);
```
Oh, I think this can be done way simpler. :) You need to wait for the process to exit and retrieve it's exit status at which point you'll know whether `find_in_proc_mount()` succeeded or not. So all the `pipe()` based logic can go. What you need is:
```
ret = c->attach(c, find_in_proc_mounts, d, &attach_options, &pid);
if (ret < 0)
        /* handle error */
ret = wait_for_pid(pid);
if (ret < 0)
        /* your function failed */

/* your function succeeded */
```

LizaTretyakova committed:
conf, confile: introduce basic structs for shared mount point

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>

LizaTretyakova committed:
conf, confile: add parsing of a shmounts config parameter

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>

LizaTretyakova committed:
utils: add shared mount point detection

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>

LizaTretyakova committed:
start: add shmount setup on container start

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>

LizaTretyakova committed:
lxccontainer: add container API function and structs for injecting a mount

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>

LizaTretyakova committed:
tests: add mount injection tests

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>

LizaTretyakova committed:
lxccontainer: add the umount API function

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>
[christian@brauner.io: minor coding-style changes]
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

LizaTretyakova committed:
tests: add tests for umount

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>

brauner committed:
tests: tweak mount injection tests

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

LizaTretyakova committed:
lxccontainer: add handling of file mounts

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>

LizaTretyakova committed:
tests: add filesystem and char device tests

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>

Commit:
lxccontainer: do_lxcapi_mount() coding-style

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
lxccontainer: reword create_mount_target()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

LizaTretyakova committed:
conf, confile, lxccontainer, start: nonfunctional changes

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>

LizaTretyakova committed:
conf, lxccontainer: fix length checks in snprintf

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>

LizaTretyakova committed:
confile: add strdup failure check

Signed-off-by: Liza Tretyakova <elizabet.tretyakova@gmail.com>
[christian.brauner@ubuntu.com: coding style]
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
include: add getgrgid_r()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

brauner committed:
Merge pull request #2406 from 2xsec/bugfix

storage: Resource leak

Commit:
coverity: #1425764

Unchecked return value

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425770

Insecure temporary file

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425771

Insecure temporary file

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425789

Unchecked return value from library

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425792

Insecure temporary file

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425793

Unchecked return value

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425799

Logically dead code

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425804

Unchecked return value

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425810

Explicit null dereferenced

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425813

Resource leak

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425818

Dereference after null check

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425819

Resource leak

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425824

Missing break in switch

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425825

Unchecked return value

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425837

String not null terminated

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425840

Unchecked return value

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425846

Unchecked return value

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425850

Insecure temporary file

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

stgraber committed:
Merge pull request #2405 from brauner/2018-04-17/getgrgid_r

configure: check for getgrgid_r()

Commit:
coverity: #1425748

Time of check time of use

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425789

Unchecked return value from library

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1425855

String not null terminated

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1437027

Read from pointer after free

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

hallyn committed:
Merge pull request #2409 from brauner/2018-06-15/coverity

coverity

2xsec committed:
secure coding: strcpy => strlcpy

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

2xsec committed:
secure coding: network: strcpy => strlcpy

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2414 from 2xsec/bugfix

secure coding: strcpy => strlcpy

2xsec committed:
secure coding: #2 strcpy => strlcpy

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

2xsec committed:
conf: remove unused argument

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2415 from 2xsec/bugfix

secure coding: #2 strcpy => strlcpy

brauner committed:
Merge pull request #2416 from 2xsec/bugfix

conf: remove unused argument

2xsec committed:
secure coding: #3 strcpy => strlcpy

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2417 from 2xsec/bugfix

secure coding: #3 strcpy => strlcpy

2xsec committed:
fix fd handle leak

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

2xsec committed:
fix pointer c is dereferenced after checking null

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2418 from 2xsec/bugfix

fix fd handle leak

2xsec committed:
secure coding: strcat => strncat

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

2xsec committed:
secure coding: cgfsng: strncat, strlcpy

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

Commit:
include: add strlcat() implementation

CC: Donghwa Jeong <dh48.jeong@samsung.com>
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

monstermunchkin committed:
templates: Fix busybox template

Use `busybox --list`, and exclude the `busybox` applet if necessary.

Signed-off-by: Thomas Hipp <thomas.hipp@canonical.com>

stgraber committed:
Merge pull request #2420 from brauner/2018-06-20/strlcat

include: add strlcat() implementation

brauner committed:
Merge pull request #2422 from monstermunchkin/fix/busybox-template

templates: Fix busybox template

Commit:
btrfs: s/strncat()/strlcat()/g

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

stgraber committed:
Merge pull request #2423 from brauner/2018-06-20/fix_arch

btrfs: s/strncat()/strlcat()/g

2xsec committed:
btrfs: fix wrong buffer size to append string

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

2xsec committed:
cgfsng: strncat => strlcat

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

2xsec committed:
conf: strncat => strlcat

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

2xsec committed:
pam_cgfs: strncat => strlcat

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

2xsec committed:
tool_utils: strncat => strlcat

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

2xsec committed:
utils: strncat => strlcat

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

2xsec committed:
confile: strncat => strlcat

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

2xsec committed:
btrfs: add to check return size of strlcat

Signed-off-by: Donghwa Jeong <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2419 from 2xsec/bugfix

secure coding: strcat => strncat

Commit:
commands: simplify lxc_cmd()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

2xsec committed:
monitor: change exit() => _exit() system call in child process

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2426 from 2xsec/bugfix

monitor: change exit() => _exit() system call in child process

stgraber committed:
Merge pull request #2425 from brauner/2018-06-22/cleanups

commands: simplify lxc_cmd()

2xsec committed:
move some comments in lxc.spec.in

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2428 from 2xsec/bugfix

move some comments in lxc.spec.in

2xsec committed:
log: add lxc_log_strerror_r macro

Let's ensure that we always use the thread-safe strerror_r() function and add
an approriate macro.
Additionally, define SYS*() macros for all log levels. They will use the new
macro and ensure thread-safe retrieval of errno values.

Signed-off-by: 2xsec <dh48.jeong@samsung.com>
[christian.brauner@ubuntu.com: simplify lxc_log_strerror_r macro]
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

brauner committed:
Merge pull request #2429 from 2xsec/bugfix

log: add strerror_r macro

duguhaotian committed:
clear ONLCR flag for container pty

now container pty output log, use \r\n as a newline flag.
This is a windows type, so we need to change it. By clear
ONLCR can reach it.

Signed-off-by: duguhaotian <duguhaotian@gmail.com>

2xsec committed:
log: change WARN macro using strerror to SYSWARN

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
log: change DEBUG, INFO, TRACE, NOTICE macro using strerror to SYS* macro

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
log: remove strerror() from SYSERROR call

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
log: change ERROR macro using sterror to SYSERROR

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2431 from 2xsec/bugfix

log: change LOG macros using strerror to SYS* macros

Commit:
log: account for Android's Bionic's strerror_r()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

brauner committed:
Merge pull request #2430 from duguhaotian/work

clear ONLCR flag for container pty

stgraber committed:
Merge pull request #2433 from brauner/2018-06-26/fix_android

log: account for Android's Bionic's strerror_r()

2xsec committed:
tools: fix quiet option is not working

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: cgroup, freeze, unfreeze, wait: change fprintf => lxc_error

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2434 from 2xsec/bugfix

tools: fix quiet option is not working

2xsec committed:
tools: lxc-start: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-stop: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-freeze: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-unfreeze: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-wait: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-cgroup: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-attach: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-autostart: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
utils: code cleanups

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-info: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-ls: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-config: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-console: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-destroy: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-device: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-execute: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-copy: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

Commit:
autodev: adapt to changes in Linux 4.18

Starting with commit
55956b59df33 ("vfs: Allow userns root to call mknod on owned filesystems.")
Linux will allow mknod() in user namespaces for userns root if CAP_MKNOD is
available.
However, these device nodes are useless since

static struct super_block *alloc_super(struct file_system_type *type, int flags,
                                       struct user_namespace *user_ns)
{
        /* <snip> */

        if (s->s_user_ns != &init_user_ns)
                s->s_iflags |= SB_I_NODEV;

        /* <snip> */
}

will set the SB_I_NODEV flag on the filesystem. When a device node created in
non-init userns is open()ed the call chain will hit:

bool may_open_dev(const struct path *path)
{
        return !(path->mnt->mnt_flags & MNT_NODEV) &&
                !(path->mnt->mnt_sb->s_iflags & SB_I_NODEV);
}

which will cause an EPERM because the device node is located on an fs
owned by non-init-userns and thus doesn't grant access to device nodes due to
SB_I_NODEV.

The solution is straightforward. Unless you're real root you should bind-mount
device nodes.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

2xsec committed:
tools: lxc-top: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-unshare: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-checkpoint: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-checkpoint: fix log name

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-snapshot: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-create: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-monitor: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: arguments: share internal API symbols

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: remote tool_list, tool_utils

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

stgraber committed:
Merge pull request #2438 from brauner/2018-06-29/linux-v4.18

autodev: adapt to changes in Linux 4.18

Commit:
conf: simplify lxc_setup_dev_console()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
terminal: safely retrieve path of slave device

openpty() is a horrible function that uses strcpy() into the char *name
argument if name != NULL. We can't rely on the path being sane in all cases so
let's split out the name retrieval to ttyname_r().

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
CODING_STYLE: add section about using strlcat()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
start: delete terminal on error

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
terminal: set FD_CLOEXEC on pty file descriptors

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>
Cc: Wolfgang Bumiller <w.bumiller@proxmox.com>

Commit:
attach: do not make terminal raw

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
tests: remove duplicate file_exists() definition

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
utils: add fd_cloexec()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>
Cc: Wolfgang Bumiller <w.bumiller@proxmox.com>

2xsec committed:
tools: lxc-autostart: remove the trailing . & fix coding style of strncmp

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-checkpoint: remove the trailing .

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-config: fix coding style of strncmp

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-copy: remove the trailing . & fix coding style of strncmp

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-create: remove the trailing . & fix coding style of strncmp

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-destroy: remove the trailing .

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-device: remove the trailing . & fix coding style of strncmp

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-snapshot: fix coding style of strncmp

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-start: remove the trailing .

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-top: fix coding style of strncmp

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-wait: remove the trailing .

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2437 from 2xsec/bugfix

tools: share internal API symbols

Commit:
tree-wide: remove unneeded log prefixes

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

stgraber committed:
Merge pull request #2440 from brauner/2018-06-30/console_fixes

terminal: security fixes

2xsec committed:
coverity: #1425816

Explicit null dereferenced

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2441 from 2xsec/coverity

coverity: #1425816

Commit:
log: cleanup

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

2xsec committed:
utils: move recursive_destroy() from cfgsng to utils.

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
sync: modify error string

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2444 from 2xsec/bugfix

utils: move recursive_destroy() from cfgsng to utils.

2xsec committed:
tests: cleanup api_reboot.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup clonetest.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup concurrent.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup config_jump_table.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup console.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup containertests.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup createtest.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup criu_check_feature.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup destroytest.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup device_add_remove.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup get_item.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup getkeys.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup list.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup locktests.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup lxc-test-utils.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup lxc_raw_clone.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup lxcpath.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: clenaup may_control.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup reboot.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup saveconfig.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup shortlived.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup shutdowntest.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup snapshot.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup startone.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tests: cleanup state_server.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2445 from 2xsec/bugfix

tests: cleanup source codes.

2xsec committed:
log: remove the trailing .

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2448 from 2xsec/bugfix

log: remove the trailing .

Commit:
cgroups: remove is_crucial_cgroup_subsystem()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
cgfsng: respect lxc.cgroup.use

If lxc.cgroup.use is specified then only those controllers listed in there will
be used others will be skipped.

Closes #2447.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
conf: use fd_cloexec()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
conf: safely retrieve path of slave device

openpty() is a horrible function that uses strcpy() into the char *name
argument if name != NULL. We can't rely on the path being sane in all cases so
let's split out the name retrieval to ttyname_r().

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
conf: use mknod() to create dummy mount target

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

stgraber committed:
Merge pull request #2446 from brauner/2018-07-02/log_prefix

Logging and cgroup fixes

Commit:
conf: use mknod() to create dummy mount targets

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
terminal: don't garble output

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

hallyn committed:
Merge pull request #2450 from brauner/2018-07-05/use_mknod

conf: use mknod() to create dummy mount targets

Commit:
start: refactor match_stdfds()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
start: declare int array const

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
tree-wide: set MSG_NOSIGNAL

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

stgraber committed:
Merge pull request #2451 from brauner/2018-07-06/bugfixes

bugfixes

Commit:
autotools: build both a shared and static liblxc

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>
Cc: Serge Hallyn <serge@hallyn.com>
Cc: Tycho Andersen <tycho@tycho.ws>

2xsec committed:
namespace: add api to convert namespaces to standard identifiers

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-attach: replace converting standard identifiers of namespaces to api

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-unshare: replace converting standard identifiers of namespaces to api

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2453 from 2xsec/bugfix

add namespace api

2xsec committed:
tools: lxc-unshare: replace write_id_mapping() to the same function in conf.c

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2454 from 2xsec/bugfix

tools: lxc-unshare: replace write_id_mapping() to the same function in conf.c

Commit:
conf: create /dev directory

If users specified lxc.autodev = 1 it does not make sense to skip setting up
autodev if /dev does not exist. We rather should create it.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
start: don't unconditionally open("/dev/null")

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

stgraber committed:
Merge pull request #2452 from brauner/2018-07-06/build_static_liblxc

static liblxc, always create /dev directory when autodev = 1, open("/dev/null") opportunistically

2xsec committed:
tools: lxc-ls: fix signed_to_bigger_unsigned issue

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

Commit:
autotools: cleanup Makefile.am

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
attach: simplify lxc_attach_getpwshell()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

brauner committed:
Merge pull request #2458 from 2xsec/bugfix

tools: lxc-ls: fix signed_to_bigger_unsigned issue

Commit:
attach: cleanup fetch_seccomp()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
attach: cleanup no_new_privs()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
attach: use fd_cloexec()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
attach: cleanup log messages in lxc_attach()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
conf: handle partially functional device nodes

This improves handling kernels which allow userspace to create partially
functional devices nodes.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

stgraber committed:
Merge pull request #2460 from brauner/2018-07-12/handle_new_mknod_smarter

conf: handle partially functional device nodes

stgraber committed:
Merge pull request #2459 from brauner/2018-07-11/cleanup_makefile

autotool fixes, attach cleanups

Commit:
coverity: #1437936

Unchecked return value

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
coverity: #1437935

Unchecked return value

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

stgraber committed:
Merge pull request #2462 from brauner/2018-07-12/coverity

coverity

2xsec committed:
lxclock: change error log using strerror to SYSERROR

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2465 from 2xsec/bugfix

lxclock: change error log using strerror to SYSERROR

flx42 committed:
confile: fix incorrect strncmp

Passing additional configuration options with "--define" was broken.

Result of git bisect:
d899f11b7bfb14c4b532bc801de89c8fb46307d4 is the first bad commit

Signed-off-by: Felix Abecassis <fabecassis@nvidia.com>

brauner committed:
Merge pull request #2466 from flx42/fix-define-argument

 confile: fix incorrect strncmp

2xsec committed:
tools: lxc-unshare: apply argument parser of lxc and log system of lxc

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
tools: lxc-unshare: fix wrong coding rules

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2468 from 2xsec/bugfix

tools: lxc-unshare: apply argument parser of lxc and log system of lxc

2xsec committed:
tools: lxc-unshare: use lxc list for interface names

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
coverity: #1437949

Argument cannot be negative

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2469 from 2xsec/bugfix

tools: lxc-unshare: use lxc list for interface names

Commit:
confile: add lxc.monitor.signal.pdeath

Set the signal to be sent to the container's init when the lxc monitor exits.
By default it is set to SIGKILL which will cause all container processes to be
killed when the lxc monitor process dies.
To ensure that containers stay alive even if lxc monitor dies set this to 0.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
confile: move signal helpers to confile utils

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

stgraber committed:
Merge pull request #2470 from brauner/2018-07-16/monitor_signal_pdeath

confile: add lxc.monitor.signal.pdeath

Commit:
tests: add lxc.monitor.signal.pdeath

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

stgraber committed:
Merge pull request #2472 from brauner/2018-07-16/monitor_signal_pdeath

tests: add lxc.monitor.signal.pdeath

tenforward committed:
doc: Translate lxc.monitor.signal.pdeath into Japanese in lxc.container.conf(5)

Signed-off-by: KATOH Yasufumi <karma@jazz.email.ne.jp>

stgraber committed:
Merge pull request #2473 from tenforward/japanese

doc: Translate lxc.monitor.signal.pdeath into Japanese in lxc.contain…

Commit:
conf: the atime flags are locked in userns

This means they need to be added for remount and for fresh mounts.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
apparmor: Allow /usr/lib* paths for mount and pivot_root

openSUSE Leap 15 is using --libdir=/usr/lib64 when building for
x86_64 so we need to allow this path in the apparmor profiles.

Link: https://bugzilla.opensuse.org/show_bug.cgi?id=1099239
Signed-off-by: Markos Chandras <mchandras@suse.de>

brauner committed:
Merge pull request #2481 from hwoarang/fix-apparmor-paths

apparmor: Allow /usr/lib* paths for mount and pivot_root

stgraber committed:
Merge pull request #2480 from brauner/2018-07-18/add_required_mount_flags

conf: the atime flags are locked in userns

2xsec committed:
confile_utils: fix return value & cleanups

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
lxccontainer: coding rules

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
af_unix: fix return value & cleanups

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2482 from 2xsec/bugfix

error handling cleanups #2471

2xsec committed:
attach: fix return value & cleanups

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
docs: add long options of lxc-unshare

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2485 from 2xsec/docs

docs: add long options of lxc-unshare

2xsec committed:
attach: move errno handling

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2484 from 2xsec/bugfix

attach: fix return value & cleanups

2xsec committed:
thread safe: rand() => rand_r()

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

2xsec committed:
coverity: #1438067

Explicit null dereferenced

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2486 from 2xsec/bugfix

thread safe: rand() => rand_r()

2xsec committed:
docs: tools: -d/--daemonize for lxc-execute

Signed-off-by: 2xsec <dh48.jeong@samsung.com>

brauner committed:
Merge pull request #2488 from 2xsec/bugfix

docs: tools: -d/--daemonize for lxc-execute

Commit:
conf: coding style fixes

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
start: coding style fixes

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
confile: add missing header

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>
